src/dbutils/mutation.rs

```diff
@@ -0,0 +1,359 @@
+type SimpleBuffer = HashMap<TableName, SimpleBucket>;
+type DupSortValues = BTreeSet<Vec<u8>>;
+
+#[derive(Default)]
+struct DupSortChanges {
+    insert: DupSortValues,
+    delete: DupSortValues,
+}
+
+type DupSortBucket = BTreeMap<Vec<u8>, DupSortChanges>;
+type DupSortBuffer = HashMap<TableName, DupSortBucket>;
+
+fn is_dupsort(table_name: &str) -> bool {
+    if let Some(info) = TABLES.get(table_name) {
+        info.dup_sort.is_some()
+    } else {
+        false
+    }
+}
+
+struct Mutation<'tx: 'm, 'm, Tx: MutableTransaction<'tx>> {
+    parent: &'m mut Tx,
+    phantom: PhantomData<&'tx i32>,
+    simple_buffer: HashMap<TableName, SimpleBucket>,
+    dupsort_buffer: HashMap<TableName, DupSortBucket>,
+    sequence_increment: HashMap<TableName, u64>,
+}
+
+impl<'tx: 'm, 'm, Tx: MutableTransaction<'tx>> Mutation<'tx, 'm, Tx> {
+    fn new(parent: &'m mut Tx) -> Self {
+        Mutation {
+            parent,
+            phantom: PhantomData,
+            simple_buffer: Default::default(),
+            dupsort_buffer: Default::default(),
+            sequence_increment: Default::default(),
+        }
+    }
+
+    async fn get<T>(&'m self, table: &T, key: &[u8]) -> Result<Option<Bytes<'m>>>
+    where
+        T: Table,
+    {
+        let table_name = table.db_name();
+
+        if let Some(bucket) = self.simple_buffer.get(&table_name) {
+            Ok(match bucket.get(key) {
+                Some(entry) => entry.as_ref().map(|value| value.as_slice().into()),
+                None => self.parent.get(table, key).await?,
+            })
+        } else if let Some(_bucket) = self.dupsort_buffer.get(&table_name) {
+            todo!()
+        } else {
+            Err(MutationError::TableNotFound.into())
+        }
+    }
+
+    async fn read_sequence<T>(&self, table: &T) -> Result<u64>
+    where
+        T: Table,
+    {
+        let parent_value = self.parent.read_sequence(table).await?;
+        let increment = self.sequence_increment.get(&table.db_name()).unwrap_or(&0);
+
+        Ok(parent_value + increment)
+    }
+
+    fn get_simple_buffer(&mut self, name: &TableName) -> &mut SimpleBucket {
+        if !self.simple_buffer.contains_key(name) {
+            self.simple_buffer.insert(name.clone(), Default::default());
+        }
+        self.simple_buffer.get_mut(name).unwrap()
+    }
+
+    fn get_dupsort_buffer(&mut self, name: &TableName) -> &mut DupSortBucket {
+        if !self.dupsort_buffer.contains_key(name) {
+            self.dupsort_buffer.insert(name.clone(), Default::default());
+        }
+        self.dupsort_buffer.get_mut(name).unwrap()
+    }
+
+    fn get_dupsort_changes(&mut self, name: &TableName, k: &[u8]) -> &mut DupSortChanges {
+        let buffer = self.get_dupsort_buffer(name);
+        if !buffer.contains_key(k) {
+            buffer.insert(k.to_vec(), Default::default());
+        }
+        buffer.get_mut(k).unwrap()
+    }
+
+    async fn set<T>(&mut self, table: &T, k: &[u8], v: &[u8]) -> Result<()>
```

Comment:
For this method and `increase_sequence()`, we would need to change the signatures in the trait if they are to implement `MutableTransaction` (`&self` to `& mut self`)

Reply:
Or use mutex. Mutex' fast path is _really_ fast - just compare an atomic.

Reply:
So `parent` will be `Arc<Mutex<T>>` instead of `&mut T`, and the same change will be made in the stage(s) using `Mutation`?